### PR TITLE
[Qwen3.5] Fix GDN linear attention multi-token cached forward

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

The gated-delta-net (GDN) forward only used the cached recurrent state when `seq_len == 1`. For any multi-token forward with a populated cache (e.g. chunked prefill continuation or speculative-decoding verification), it fell through to `chunk_gated_delta_rule(initial_state=None)`, silently restarting the linear layers from zero and ignoring the prefill state.

This breaks the causal-LM invariant that the logits at position `i` must not depend on whether later tokens are batched into the same call — position 0 of a 16-token verify forward ended up differing from the corresponding single-token cached decode, collapsing to high-frequency context tokens and destroying speculative-decoding correctness.

Add a `use_cached_chunk` path that, when `has_previous_state` is true and `seq_len > 1`:
- reads the cached `conv_state` / `recurrent_state`,
- prepends the conv context onto the chunk input so the causal conv sees the correct left-context,
- drops the prepended context from the output,
- passes the cached `recurrent_state` as `initial_state` to `chunk_gated_delta_rule`.

The same fix propagates to `qwen3_5_moe` via the modular system and is applied to `qwen3_next` and `olmo_hybrid`.

Add a unit test that compares the first-position output of a multi-token cached forward against the single-token cached forward on the same token and cache. Without this fix the mismatch is 100%.

Fixes # (issue)

## Code Agent Policy

The Transformers repo is currently being overwhelmed by a large number of PRs and issue comments written by
code agents. We are currently bottlenecked by our ability to review and respond to them. As a result, 
**we ask that new users do not submit pure code agent PRs** at this time. 
You may use code agents in drafting or to help you diagnose issues. We'd also ask autonomous "OpenClaw"-like agents
not to open any PRs or issues for the moment.

PRs that appear to be fully agent-written will probably be closed without review, and we may block users who do this
repeatedly or maliciously. 

This is a rapidly-evolving situation that's causing significant shockwaves in the open-source community. As a result, 
this policy is likely to be updated regularly in the near future. For more information, please read [`CONTRIBUTING.md`](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md).

- [ ] I confirm that this is not a pure code agent PR.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by Cyrilvallez** [comment]
cc @vasqu here to double-check as well: I had raised similar concerns when I was refactoring the mamba caches, but I'm less familiar than you with all the mamba kernels and how the updates should exactly look like. But some mambas use the `seq_len == 1` part and other don't, and as I said before idk if it was really  intended or not.

Fix looks legit though, but would appreciate second opiniuon!

**Comment by kashif** [comment]
Thanks for the thorough check! The `seq_len == 1` pattern isn't intentional — I believe it was copied forward from mamba → qwen3_next → qwen3_5 without anyone  exercising the "populated cache + seq_len > 1" path (plain `.generate()` always has `seq_len == 1` in decode, and training doesn't use the cache). I grepped  transformers and the same `has_previous_state(...) and seq_len == 1` guard appears in 9 mamba-family files: `bamba`, `falcon_h1`, `granitemoehybrid`, `jamba`,  `olmo_hybrid`, `qwen3_next`, `qwen3_5`, `qwen3_5_moe`, `zamba`. Every one of them drops the cached recurrent state in the chunk path (passes  `initial_state=None`) and zero-pads the conv state instead of prepending the cached context.

The canonical reference is `flash-linear-attention`. I grepped `fla/layers/` and every linear-attention implementation (`gated_deltanet`, `delta_net`, `gla`,  `simple_gla`, `hgrn`, `hgrn2`, `rwkv6`, `rwkv7`, `lightnet`, `gated_deltaproduct` — 10 I checked) follows the same canonical pattern:

  ```python
  recurrent_state = last_state['recurrent_state'] if last_state is not None else None
  mode = 'fused_recurrent' if q_len <= 64 else self.mode   # perf heuristic, not correctness
  o, recurrent_state = chunk_or_fused_recurrent_kernel(..., initial_state=recurrent_state, ...)
```
  No seq_len gating on correctness. The `q_len <= 64` cutoff in FLA is a kernel-selection heuristic (fused-recurrent is faster for short seqs), not a correctness   gate — both kernels accept and propagate `initial_state`.

  This fix brings qwen3_5 in line with FLA. I've constrained the change to qwen3_5 + qwen3_5_moe (via modular) so it's reviewable; happy to open follow-up PRs  for the other mamba-family files if you'd like them audited together.

  The new unit test (`test_linear_attention_multi_token_cached_forward_matches_single_token`) is the minimal invariant the old code violated — a causal LM's logits   at position `i` must not depend on whether later tokens are batched into the same call, across any combination of `seq_len` and cache state. 100% element mismatch  on main, passes after the fix.

File: src/transformers/models/qwen3_5/modular_qwen3_5.py [hunk 1]
```diff
@@ -218,9 +218,15 @@ def forward(
         use_precomputed_states = (
             cache_params is not None and cache_params.has_previous_state(self.layer_idx) and seq_len == 1
         )
+        # Multi-token forward after a prior forward has populated the cache (e.g. chunked prefill
+        # continuation or speculative verification). The single-token fused kernel can't be used
+        # here, so we fall into the chunked kernel and must carry forward both the cached conv
+        # context and the cached recurrent state; otherwise the first-position outputs silently
+        # diverge from the equivalent single-token decode.
+        use_cached_chunk = cache_params is not None and cache_params.has_previous_state(self.layer_idx) and seq_len > 1
```
Comment (vasqu) [inline src/transformers/models/qwen3_5/modular_qwen3_5.py]: Why not remove the `seq_len > 1` then? I agree that this shouldn't be a restriction as it will be used as initial state and the forward essentially stays the same. Same/Similar for qwen3 next then

We have something similar to Mamba2, unsure about Mamba1. Would be a nice to have there as well but not directly needed in this PR. Reminds me of these lines that were removed https://github.com/huggingface/transformers/blob/83a6c5b577cafa607d59e78af4de86592b9903ee/src/transformers/models/mamba2/modeling_mamba2.py#L642-L645

File: src/transformers/models/qwen3_5/modular_qwen3_5.py [hunk 2]
```diff
@@ -244,9 +250,13 @@ def forward(
                 self.activation,
             )
         else:
+            if use_cached_chunk:
+                # Prepend the cached conv context so the causal conv sees the correct left-context
+                # when continuing a prior forward rather than starting from zero-padding.
+                mixed_qkv = torch.cat([conv_state, mixed_qkv], dim=-1)
```
Comment (vasqu) [inline src/transformers/models/qwen3_5/modular_qwen3_5.py]: Ok I see now, we have to accomodate for the seq_len > 1 case. I'd still keep the terms `precomputed_state` and something along `single_token vs chunk_tokens` to have clearer boundaries

**Review by vasqu — APPROVED** [review]
Happy to have that, it is a valid usecase and just a limitation from our side tbh (which you lift here). I feel like something similar could be done for Mamba2 as well but maybe out of scope for here

My main comments is more about conventions and maybe propagating to qwen3 next and olmo hybrid if possible

**Comment by kashif** [comment]
thanks @vasqu let me simplify and update

**Comment by kashif** [comment]
should i do the qwen3_next and olmo_hybrid fixes in a followup PR or this one @vasqu ?

File: src/transformers/models/qwen3_5/modular_qwen3_5.py [hunk 3]
```diff
@@ -215,12 +215,20 @@ def forward(
...
+        #   * single-token (`use_precomputed_states`): fused per-step kernel — standard `.generate()`.
+        #   * chunk-tokens (`use_cached_chunk`): multi-token forward after the cache was populated
+        #     (chunked-prefill continuation / speculative-decoding verification). The fused single-step
+        #     kernel can't be used; fall into `chunk_gated_delta_rule`, but still carry forward both
+        #     the cached conv context and the cached recurrent state — otherwise the first-position
+        #     outputs silently diverge from the single-token path.
         use_precomputed_states = (
             cache_params is not None and cache_params.has_previous_state(self.layer_idx) and seq_len == 1
         )
+        use_cached_chunk = cache_params is not None and cache_params.has_previous_state(self.layer_idx)
```
Comment (vasqu) [inline src/transformers/models/qwen3_5/modular_qwen3_5.py]:
```suggestion
```
Ah sorry I meant it a bit differently, imo both are precomputed states that are differing in their amount of tokens so I'd like to split the logic explicitly with focusing on seq_len where needed. So
1. Only use precomputed states (and remove the seq len condition)
2. Where needed check for the initial seq len to determine if we have a single decode step or chunks

**Comment by vasqu** [comment]
@kashif would be nice if we could do those models here as well - they are all based on GDN

**Comment by kashif** [comment]
yup! fixing all GDN ones

**Comment by kashif** [comment]
@vasqu if you can have another look whenever you are free

File: src/transformers/models/olmo_hybrid/modular_olmo_hybrid.py [hunk 4]
```diff
@@ -342,10 +342,16 @@ def forward(
             )
             conv_state = x_with_state[:, :, 1:]
         else:
-            # Multi-token forward (prefill mode)
```
Comment (vasqu) [inline src/transformers/models/olmo_hybrid/modular_olmo_hybrid.py:345]: I think we can keep the comment and adjust `prefill mode/multi-token decode`

File: src/transformers/models/qwen3_5/modular_qwen3_5.py [hunk 5]
```diff
@@ -256,7 +262,10 @@ def forward(
                     seq_idx=None,
                 )
             else:
-                mixed_qkv = F.silu(self.conv1d(mixed_qkv)[:, :, :seq_len])
+                mixed_qkv = F.silu(self.conv1d(mixed_qkv)[:, :, : mixed_qkv.shape[-1]])
+            if use_precomputed_states:
+                # Drop the prepended context; only this chunk's outputs remain.
```
Comment (vasqu) [inline src/transformers/models/qwen3_5/modular_qwen3_5.py]: Let's add this comment to the first if block related to this - similar to olmo hybrid --> prepend context and drop it afterwards

**Review by vasqu — APPROVED** [review]
LGTM, just some last nits left, then we are good to merge 🤗

**Comment by vasqu** [comment]
run-slow: olmo_hybrid, qwen3_5, qwen3_5_moe, qwen3_next

**Comment by vasqu** [comment]
Just sanity checking with run-slow, but should be good will merge afterwards

**Comment by vasqu** [comment]
@kashif thanks for the PR :hugs: merging now!
